drivers/spi/spi_emul.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * This driver creates fake SPI buses which can contain emulated devices,
+ * implemented by a separate emulation driver. The API between this driver and
+ * its emulators is defined by struct spi_emul_driver_api.
+ *
+ * Copyright 2020 Google LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_spi_emul_controller
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(spi_emul_ctlr);
+
+#include <device.h>
+#include <emul.h>
+#include <drivers/spi.h>
+#include <drivers/spi_emul.h>
+
+/** Working data for the device */
+struct spi_emul_data {
+	/* List of struct spi_emul associated with the device */
+	sys_slist_t emuls;
+	/* SPI host configuration */
+	uint32_t config;
+};
+
+uint32_t spi_emul_get_config(const struct device *dev)
+{
+	struct spi_emul_data *data = dev->data;
+
+	return data->config;
+}
+
+/**
+ * Find an emulator for a SPI bus
+ *
+ * At present only a single emulator is supported on the bus, since we do not
+ * support chip selects.
+ *
+ * @param dev SPI emulation controller device
+ * @return emulator to use
+ * @return NULL if not found
+ */
+static struct spi_emul *spi_emul_find(const struct device *dev)
+{
+	struct spi_emul_data *data = dev->data;
+	sys_snode_t *node;
+
+	SYS_SLIST_FOR_EACH_NODE(&data->emuls, node) {
+		struct spi_emul *emul;
+
+		emul = CONTAINER_OF(node, struct spi_emul, node);
+		return emul;
+	}
+
+	return NULL;
+}
+
+static int spi_emul_io(const struct device *dev,
+		       const struct spi_config *config,
+		       const struct spi_buf_set *tx_bufs,
+		       const struct spi_buf_set *rx_bufs)
+{
+	struct spi_emul *emul;
+	const struct spi_emul_api *api;
+	int ret;
+
+	emul = spi_emul_find(dev);
+	if (!emul) {
+		return -EIO;
```

Comment:
Maybe, but it seems a bit odd to return that there is no device:

- clearly @dev is a device
- the emulator is (for now at least) not a device

This is akin to the hardware being disconnected, so it simply doesn't respond. Perhaps a better emulation would be to return 0 and do nothing? But I worry that it would be hard for people to debug.

What do you think?

Reply:
Ah, I see that more clearly now. I think `-EIO` makes sense